prune_shard_list.c

```diff
@@ -239,6 +249,60 @@ MakeOpExpression(Var *variable, int16 strategyNumber)
 
 
 /*
+ * LookupOperatorByType is a wrapper around GetOperatorByType that uses a cache
+ * to avoid multiple lookups of operators within a single session by their types.
+ */
+static Oid
+LookupOperatorByType(Oid typeId, Oid accessMethodId, int16 strategyNumber)
+{
+	OperatorIdCacheEntry *matchingCacheEntry = NULL;
+	ListCell *cacheEntryCell = NULL;
+
+	/* search the cache */
+	foreach(cacheEntryCell, OperatorIdCache)
+	{
+		OperatorIdCacheEntry *cacheEntry = lfirst(cacheEntryCell);
+
+		if ((cacheEntry->typeId == typeId) &&
+			(cacheEntry->accessMethodId == accessMethodId) &&
+			(cacheEntry->strategyNumber == strategyNumber))
+		{
+			matchingCacheEntry = cacheEntry;
+			break;
+		}
+	}
+
+	/* if not found in the cache, call GetOperatorByType and put the result in cache */
+	if (matchingCacheEntry == NULL)
+	{
+		Oid operatorId;
```

Comment:
Here I think it's actually best to reverse the order of the variables and set `operatorId` immediately to the return of `GetOperatorByType`.